### DISCUSSION TITLE
Ideas to improve performance for dbachecks with Pester v5 compared to v4? Getting a single Tag from a file with 100 +Tags

### DISCUSSION THREAD

**Discussion by SQLDBAWithABeard** [pester/Pester#2331] [Q&A] [answered]
Terminology - a `check` when stated below can be thought of as a Pester Describe block with a unique Tag name which contains a single Context and 1 or more It blocks.

We are currently rewriting [dbachecks ](https://github.com/dataplat/dbachecks) to use Pester v5 (using this [project ](https://github.com/orgs/dataplat/projects/2) in case anyone fancies helping out! )

So that we can ensure that when we rewrite we are testing the same, passing for the same and failing for the same because these are databases we are talking about and we are (ex) DBAs and cautious folk, @jpomfret and I wrote a dev container configuration with 3 sql instances and also a function that is loaded into the session that will run the same check against the 3 instances in the dev container and compare the output to see if they are the same. (This has caught some coding misconceptions :-) ) We also use Sampler to build and deploy the module so that we are developing in clean sessions ([ if you want to follow along](https://github.com/dataplat/dbachecks/blob/main/developing/Howto.md) instructions are here)

We also want to improve the performance of the checks as we had scenarios where you could run 5 different checks against the same instance and actually return to the instance 5 times to get the same information !! So we added to the above function to use Jakubs profier module to work out what is happening.

Now we want to gather the information from a single trip to the instance so [we identify the unique checks that we are running, work out which files they are in](https://github.com/dataplat/dbachecks/blob/c0b12ed7d3d2fe31ccca6676ef7f03633ee75149/source/internal/functions/Invoke-DbcCheckv5.ps1#L148) and add that to the `$Configuration.Run.Path` property.

So that is the background. Now the issue.

If we look at a number of checks like this

```
$Checks = 'ErrorLogCount', 'XESessionRunning', 'XESessionRunningAllowed', 'XESessionExists', 'XESessionStopped', 'XpCmdShellDisabled', 'WhoIsActiveInstalled', 'CLREnabled', 'TwoDigitYearCutoff', 'MaxDopInstance', 'ErrorLogCount', 'ModelDbGrowth', 'DefaultBackupCompression', 'SaExist', 'SaDisabled', 'SaRenamed', 'DefaultFilePath', 'AdHocDistributedQueriesEnabled', 'AdHocWorkload', 'DefaultTrace', 'OleAutomationProceduresDisabled', 'CrossDBOwnershipChaining', 'ScanForStartupProceduresDisabled', 'RemoteAccessDisabled', 'SQLMailXPsDisabled', 'DAC', 'OLEAutomation', 'ServerNameMatch', 'OrphanedFile', 'MaxMemory', 'NetworkLatency'
# and run the validation
Invoke-PerfAndValidateCheck -Checks $Checks
```
then it runs for a minute and outputs

![image](https://user-images.githubusercontent.com/6729780/228861804-c3b5a784-cece-441f-8118-6d9c30667bae.png)

We are amazing, running the same tests against the same instances on the same hardware gives the same results and runs in 50% of the time with our new code. AWESOME.

However, and the reason for this question. If we change that to run only a single check

```
 $Checks = 'OrphanedFile'
# and run the validation
Invoke-PerfAndValidateCheck -Checks $Checks
```

We get

![image](https://user-images.githubusercontent.com/6729780/228862535-67ee5b4a-940b-4aa7-8459-5d1c1eeba90d.png)

Yes with different single checks, yes the hardware is the same, we are in a dev container, the test run straight after each other, it doesnt matter the host that you use, you can even get the same resutls in the codespace. So we can look at the perf detail. We normally use this to try and improve our code.

Running a single check takes between 150% and 300% longer

```
 $Checks = 'OrphanedFile'
# and run the validation
Invoke-PerfAndValidateCheck -Checks $Checks -PerfDetail
```
![image](https://user-images.githubusercontent.com/6729780/228863803-f115fbac-611c-4e52-aa0e-c67c3d2a51d9.png)

So we have something using 40+% of our duration, lets find out what

```
$SQLInstances = $dbachecks1, $dbachecks2, $dbachecks3 = 'dbachecks1', 'dbachecks2', 'dbachecks3'
$password = ConvertTo-SecureString "dbatools.IO" -AsPlainText -Force
$cred = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList "sqladmin", $password
$show = 'All'

$traci = Trace-Script -ScriptBlock {
    $v5code = Invoke-DbcCheck -SqlInstance $Sqlinstances -SqlCredential $cred -Check $Checks -legacy $false -Show $show -PassThru
}
$traci.Top50SelfDuration 
```
$traci.Top50SelfDuration[0] |fl                                                                                                                                                      

```
Percent      : 37.878Duration     : 00:00:04.6019035                                                                          
SelfPercent  : 37.878
SelfDuration : 00:00:04.6019035
HitCount     : 38
File         : Pester.psm1
Line         : 12058
Function     : 
Module       : dbachecks
Text         : foreach ($bootstrapFunction in (& $Get_Command -Name "PesterMock_*")) {
Path         : /workspace/output/RequiredModules/Pester/5.3.3/Pester.psm1
Hits         : {00000000-0000-0000-0000-000000000000, 00000000-0000-0000-0000-000000000000, 00000000-0000-0000-0000-000000000000, 00000000-0000-0000-0000-000000000000…}
CommandHits  : {[40, Profiler.CommandHit: Line=12058; Column=40; HitCount=19; SelfDuration=00:00:04.6015251; Text='(& $Get_Command -Name "PesterMock_*")'], [18, Profiler.CommandHit: 
               Line=12058; Column=18; HitCount=19; SelfDuration=00:00:00.0003784; Text='$bootstrapFunction']}
```
So it appears that it is this code that is causing the slowup, which I don't claim to understand the reasoning for.
```
        foreach ($bootstrapFunction in (& $Get_Command -Name "PesterMock_*")) {
            & $Remove_Item "function:/$($bootstrapFunction.Name)"
        }
 ```

The question is - how do we fix this? 
How do we get teh same performance improvements for 1 check (Tag) as we do for many?

Answers on a post card please.

Thank you
Rob - @SqlDbaWithABeard
Jess - @jpomfret
Cláudio - @ClaudioESSilva
Shane - @shaneis

**Comment by nohwnd** [comment, marked as answer]
@SQLDBAWithABeard  that code is cleaning up orphaned mock hook functions that might be in place when you cancel a run. In non-interactive session this code should be rarely needed (so maybe we can skip it). It runs once per run (once per Invoke-Pester), how many times are you calling Invoke-Pester? It shows 38 hits above. The time impact is probably constant ~5s for both multiple filters and single filter.

**Reply by SQLDBAWithABeard** [reply]
Invoke-Pester is running once Invoke-DbcCheck is a wrapper over Invoke-Pester.

It is taking longer for a single tag than it is for a whole bunch.

Its an excellent point though.

With multiple checks we get

![image](https://user-images.githubusercontent.com/6729780/228878965-a21defe7-3670-4e21-8538-7ce42a590886.png)

With a single check we get

![image](https://user-images.githubusercontent.com/6729780/228879276-3b6a4c9e-8396-4341-812d-3fc80f18cd5c.png)


With v4 Pester and multiple checks we get
![image](https://user-images.githubusercontent.com/6729780/228879944-1184b45a-7996-4029-8be7-723d21394d6d.png)

and v4 and a single check 

![image](https://user-images.githubusercontent.com/6729780/228880160-cf3079af-efc3-462a-a120-17b578515598.png)


The difference between 50% runtime for many checks and greater than 100% run time for 1 or 2 is constant

**Reply by nohwnd** [reply]
> It is taking longer for a single tag than it is for a whole bunch.

I see roughly the same duration in the first two screenshots. 3+ seconds. The percentage is different, but that is because with fewer checks the total time is shorter.

I don't see it at all in pester v4 output, am I just blind?

**Reply by SQLDBAWithABeard** [reply]
Nope, I dont see it either and yes the duration is roughly the same, this is whay I have started this discussion because I cant see why a single check is so much worse performing comparing v4 to v5 than multiple

**Reply by nohwnd** [reply]
It is there in v4, at least in the last screenshot. 5th line, takes 0.17 seconds.

**Reply by nohwnd** [reply]
That aside the second slowest command is reading the json file. You should remove out-string, and use -Raw with Get-Content, or if you need to keep compatibility IO.File.ReadAllText

**Reply by nohwnd** [reply]
Difference for the mock removal code is that you get 2 hits in v4 vs 40 in v5. 

Pester 4 does not check in all loaded modules, v5 does. https://github.com/pester/Pester/blob/4.10.1/Functions/Mock.ps1#L1589-L1604

**Reply by SQLDBAWithABeard** [reply]
You are correct, there are more improvements that we can make.

Are we saying that for a single check (tag) it will always be slower because of the impact of this remove function?

Not asking to change it, just to understand, so we can tell people. When dbachecks is used against thousands of instances it will make a difference :-)

**Reply by SQLDBAWithABeard** [reply]
Notice also, that you still get those 2 hits in v5, they are further down - you also get the 40 hits at the top ;-)

**Reply by nohwnd** [reply]
If I am not wrong, I think removing the aliases is enough. Removing the functions is extra work that does not make the tests any more correct, because without the alias there is no mocking going on, and having one extra function in the module is not hurting anything. The real bottleneck here is Get-Command, we could probably grab functions for the current session state faster, but if we entirely skip getting them then it makes it faster for everyone. Get-Alias does not seem to show up in the table, so it probably is much faster.

Try removing those two foreaches from Pester code, and see what is the result.

https://github.com/pester/Pester/blob/main/src/functions/Mock.ps1#L1590-L1592

https://github.com/pester/Pester/blob/main/src/functions/Mock.ps1#L1600-L1602

**Reply by nohwnd** [reply]
> Notice also, that you still get those 2 hits in v5, they are further down - you also get the 40 hits at the top ;-)

Yup, Pester 4 does just that. Pester 5 also does the same thing in every single loaded module, in case you created a mock in a module, and then you cancelled.

**Reply by SQLDBAWithABeard** [reply]
Yes, that has made a difference. Brought a single check up to near parity which is ok

![image](https://user-images.githubusercontent.com/6729780/228892106-d2d0a726-e3df-4417-81cc-e9af8780f6a4.png)

**Reply by SQLDBAWithABeard** [reply]
Without commenting it out I can also improve by doing this
```
$Modules = 'dbatools', 'dbachecks', 'PsFramework','Pester','JessAndBeard'
Get-Module | Where Name -notin $modules | Remove-Module -Force
```
Not sure that that is a viable option for production though !!!!

**Reply by nohwnd** [reply]
Not removing the functions should be totally fine for production. What is the difference now? How much time do you need to shave off of those top 3 commands?

**Reply by SQLDBAWithABeard** [reply]
Now we are at about a second. But as each check uses different dbatools or SMO code (we have optimised our SMO usage) we can approach those during our conversion for each check.  We can do more of that now that I am not chasing around wondering what was going on!!

Makes more sense too as it has got more noticeable since we moved to using Sampler which has increased the number of modules that we have loaded

**Reply by nohwnd** [reply]
PR here: https://github.com/pester/Pester/pull/2332

**Comment by nohwnd** [comment]
Could you also run `(Measure-Command { 1..38 | % { Get-Command -Name "PesterMock_*" }}).TotalSeconds` on the server? Shows 1 second for me locally, and you probably have way more cmdlets imported.

**Comment by SQLDBAWithABeard** [comment]
In the container on my main machine we get 6.4
![image](https://user-images.githubusercontent.com/6729780/228880820-56a70a53-2380-4218-8485-c2bd3921d50a.png)

**Comment by SQLDBAWithABeard** [comment]
and then I find another one !! - https://github.com/pester/Pester/issues/2335
